src/visitors/global_var_visitor.cpp

```diff
@@ -0,0 +1,65 @@
+/*************************************************************************
+ * Copyright (C) 2018-2019 Blue Brain Project
+ *
+ * This file is part of NMODL distributed under the terms of the GNU
+ * Lesser General Public License. See top-level LICENSE file for details.
+ *************************************************************************/
+
+#include <iostream>
+#include <memory>
+
+#include "ast/global.hpp"
+#include "ast/neuron_block.hpp"
+#include "ast/program.hpp"
+#include "ast/range.hpp"
+#include "ast/statement_block.hpp"
+#include "visitors/global_var_visitor.hpp"
+
+namespace nmodl {
+namespace visitor {
+
+void GlobalToRangeVisitor::visit_neuron_block(ast::NeuronBlock& node) {
+    ast::RangeVarVector range_variables;
+    std::vector<std::string> global_variables_to_remove;
+
+    auto& statement_block = node.get_statement_block();
+    auto& statements = (*statement_block).get_statements();
+    const auto& symbol_table = ast->get_symbol_table();
+
+    for (auto& statement: statements) {
+        /// only process global statements
+        if (statement->is_global()) {
+            const auto& global_variables =
+                std::static_pointer_cast<ast::Global>(statement)->get_variables();
+            global_variables_to_remove.clear();
+            for (auto& global_variable: global_variables) {
+                auto variable_name = global_variable->get_node_name();
+                /// check if global variable is being updated in the mod file
+                if (symbol_table->lookup(variable_name)->get_write_count() > 0) {
+                    range_variables.emplace_back(new ast::RangeVar(global_variable->get_name()));
+                    global_variables_to_remove.emplace_back(variable_name);
+                }
+            }
+
+            /// remove offending global variables
+            for (auto& global_var_to_remove: global_variables_to_remove) {
+                const auto& global_var_iter =
+                    find_if(global_variables.begin(),
+                            global_variables.end(),
+                            [&](const std::shared_ptr<ast::GlobalVar>& var) {
+                                return var->get_node_name() == global_var_to_remove;
+                            });
+                std::static_pointer_cast<ast::Global>(statement)->erase_global_var(global_var_iter);
+            }
```

Comment:
```suggestion
            for (auto& global_var_to_remove: global_variables_to_remove) {
                std::remove_if(global_variables.begin(), global_variables.end(),
                    [&](const std::shared_ptr<ast::GlobalVar>& var) {
                                return var->get_node_name() == global_var_to_remove;
                            });
            }
```
Wouldn't this be simpler and cleaner?

Reply:
It's true that it would be nicer in this case to directly remove the variable from the `globar_variables` but the `nmodl` API doesn't support this. This is something I also discussed with @cattabiani and he pointed me to the `erase_<node>` function to do this task, which is better to avoid deleting stuff without updating the parent

Reply:
Since you're simply deleting the global vars, do you need to update the parent pointer? or am I missing something?

Reply:
`parent pointer` is up-link. When you remove the node itself, you don't need to update? (cc: @cattabiani )

Reply:
I checked again and `erase_global_var` simply erases the variable from the `GlobalVarVector`, however in `std::static_pointer_cast<ast::Global>(statement)->get_variables()` `get_variables()` is a `const`, so I believe that it's not possible to delete any variable outside `erase_global_var`

Reply:
So, about erase there is the routine to erase stuff: erase_*. It was difficult to block people from operating directly on the children (adding a child requires an update of the parent pointer) without preventing them to also erase the node. I agree that is erase is safe. However, I do not see an easy way to allow direct access to the node ONLY for deletion.

Reply:
I do not really understand your point @iomaganaris . The idea is to do:
- A = get_variables()
- loop over A filling a container with the stuff that you want to erase
- call erase_global_var to erase them all at once

PS: get_variables returns a const& so it should be always up to date

Reply:
Ok, sorry but I read things in the wrong order. @ohm314 's solution is good but it cannot work here because stuff is const &. However...

You could take a look to:

https://github.com/BlueBrain/nmodl/blob/7c2c6903a829489148b414a5121c06dc4c338e06/src/visitors/visitor_utils.cpp#L126 

it does exactly the thing you want to do for statements. You can easily adapt it to globalvar. Even better, we could create a template that does this job.